### DISCUSSION TITLE
changes in iterators

### DISCUSSION THREAD

**Discussion by igotfr** [carbon-language/carbon-lang#1670] [Language design]
```c
whl (cond) {}
```
```c++
itr (e <- a) {}
```
instead `for (e in a)`
`for` would be used only as c-like `for`
`in` would be used to check if an element is `in` a collection, for instance:
```python
2 in [1, 2, 3]
```
```rust
for (let i: u8 = 0; i < a.len; i++) {}
```

**Comment by lexi-nadia** [comment]
C-style for-loops are complex to read and write, especially in cases where the index isn't just an integer. That's why Carbon (and Swift, Go, Kotlin, Rust...) use `for` for logical iteration. Any C-style for-loop can be easily expressed as a `while` loop, and the result is less "magic" (it's more obvious when each line is executed):

```swift
let i: u8 = 0;
while (i < a.len) {
  ...
  i++;
}
```

But also, this type of loop should be a lot less common in Carbon than in C++, due to things like slices.

**Reply by igotfr** [reply]
> C-style for-loops are complex to read and write, especially in cases where the index isn't just an integer. That's why Carbon (and Swift, Go, Kotlin, Rust...) use `for` for logical iteration. Any C-style for-loop can be easily expressed as a `while` loop, and the result is less "magic" (it's more obvious when each line is executed):
> 
> ```swift
> let i: u8 = 0;
> while (i < a.len) {
>   ...
>   i++;
> }
> ```
> 
> But also, this type of loop should be a lot less common in Carbon than in C++, due to things like slices.

Carbon may not have C-style `for`, even so, `itr` is more idiomatic
Rust uses the macro matches! instead `in`, just because `in` is reserved for use in `for`, <- would solve it

**Comment by lexi-nadia** [comment]
Please feel free to write up a formal proposal and send it for comment. :)

**Comment by tkoeppe** [comment]
Please also see #471, which proposed some concrete details of loops, but was deferred due to the general lack of an overall loop design.
